modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj

```diff
@@ -217,6 +217,20 @@
   [x]
   (:bigquery-cloud-sdk/temporal-type (meta x)))
 
+(def ^:private temporal-type->supported-units
+  {:timestamp #{:microsecond :millisecond :second :minute :hour :day}
+   :datetime  #{:microsecond :millisecond :second :minute :hour :day :week :month :quarter :year}
+   :date      #{:day :week :month :quarter :year}
+   :time      #{:microsecond :millisecond :second :minute :hour}})
+
+(defmethod temporal-type :+
+  [[_ f [_interval _amount unit]]]
+  (when-let [field-type (temporal-type f)]
+    (if (and (= :timestamp field-type)
+             (not (contains? (temporal-type->supported-units :timestamp) unit)))
+      :datetime
+      field-type)))
```

Comment:
This seems a little bit wonky it at least needs a comment explaining in detail why we're doing this. Otherwise if we come back and look at this in a few months it's going to be super confusing

Reply:
At any rate this isn't how I'd go about doing this

This is the code that takes intervals and generates either `timestamp_add`, `datetime_add`, etc. is at

https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L627-L652

It seems like if you wanted to support doing something like adding a `:month` to a `:timestamp` you could just replace that `assert-addable-unit` stuff. Maybe something like this

```clj
(if (or (not add-fn)
        (not (addable-unit? t unit)))
  (recur (->temporal-type :datetime hsql-form))
  (hformat/to-sql (hsql/call add-fn hsql-form (interval amount unit))))
```

Reply:
Oh nevermind. We're already actually doing almost the same thing you did here right here:

https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L662-L670

If you did something like `<timestamp> + [:interval -1 :month]` then it should give you `cast(<timestamp> AS datetime) + [:interval -1 :month]` which would ultimately compile to a `datetime_add()` and already have `temporal-type` of `:datetime`. So I'm not sure I understand what your change does. It should already work that way. I would suggest writing tests specifically for the `temporal-type` itself so it's clear what behavior you're changing here. (We also have tons of tests around `temporal-type` already, starting at https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk/query_processor_test.clj#L274, and I would be doing anything to it without adding more tests.)

Reply:
The change fixes the bug by forcing the reconciliation of temporal units. The tweaked version of your test reproduces the problem. Are you suggesting I should omit the test cases that passed without the fix? (I kept it in order to increase my confidence that I don't break something here.)

Reply:
Oh ok I think I understand what's happening. `reconcile-temporal-types` is getting called here

https://github.com/metabase/metabase/blob/master/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L604-L609

before we actually take the `:+` and call `add-interval-honeysql-form` to compile it, which happens here:

https://github.com/metabase/metabase/blob/master/src/metabase/driver/sql/query_processor.clj#L416-L424

So maybe the right to fix this is to move the call to `reconcile-temporal-types` to happen after the call to the parent compilation method instead of before that call. I don't think that would break anything but you'll have to try it out to be sure. Then you don't have to duplicate any logic anywhere, you just need to change

```clj
((get-method sql.qp/->honeysql [:sql filter-type])
 driver
 (reconcile-temporal-types clause))
```

to

```clj
(reconcile-temporal-types
 ((get-method sql.qp/->honeysql [:sql filter-type])
  driver
  clause))
```

That's what I'd try doing. Make sense?

> Are you suggesting I should omit the test cases that passed without the fix?

No.

Reply:
> That's what I'd try doing. Make sense?

Absolutely. I assumed there was some deep reason why it was not like that in the first place. The swap didn't work "out-of-the-box", but I can have a look if I can make it work or if I just did something wrong.

Reply:
@camsaul Just flipping the order of `sql.qp/->honeysql` and `reconcile-temporal-types` doesn't work. It breaks because it generates an `AddIntervalForm` which tries to add years to a timestamp. If I follow the code right, what happens is that
1. an `AddIntervalForm` w/o fixed temporal type is generated from the `[:relative-datetime -7 :year]` expression.
1. the field is converted to an `AddIntervalForm` with `:timestamp` as a fixed type,
1. `reconcile-temporal-types` forces all types to `:timestamp` (because that's the first temporal type it sees),
1. forcing the expression generated in step 1. to `:timestamp` generates no cast, but simply registers `:timestamp` as the type.
I found following the code quite difficult, so I might have misread how exactly we get this result, but I'm pretty sure that the problem is the invalid addition I mentioned.

What is even more difficult is judging what the best fix is, because I can only see what _is_ happening but cannot tell what _should_ be happing, what the intention was, etc.

What do you suggest? Go down this road further and fix the invalid `AddIntervalForm`?

Reply:
> I found following the code quite difficult

Don't hesitate to rework it then if you have a way to make it clearer. We're a big 👍on constantly improving the code

> What do you suggest?

I played around with this REPL and figured it out. The issue is that the `AddIntervalForm` might be returning a `datetime` if it ends up using `datetime_add()` because the unit isn't allowed for `timestamp` (e.g. `month`), however `(temporal-type <AddIntervalForm>)` still returns `:timestamp` in this case because it hasn't done the switch to using `datetimes` yet (it doesn't happen until the call to `to-sql`).

So the fix here is to do that switch from a `timestamp` to a `datetime` right away instead of delaying it until `to-sql` compilation. That will mean `temporal-type` will return the right thing.

The second issue is that `->temporal-type` didn't do the right thing either. If you did

```clj
(->temporal-type :timestamp <AddIntervalForm>)
```

it would still produce something that's actually a `:datetime` if the unit wasn't allowed for `:timestamp`, since it attempted to update the underlying HoneySQL form rather than do any casting. The fix here is just to detect this specific case and add a cast to make sure we get the right type. 

I put together https://github.com/metabase/metabase/pull/23367 which demonstrates the changes I made to get things working correctly. There are also a couple of functions in there, `(x)` and `(y)` you can use for testing it from the REPL.